server/remote_cache/cachetools/cachetools.go

```diff
@@ -1109,20 +1109,32 @@ func getAndCacheTreeFromRootDirectoryDigest(ctx context.Context, casClient repb.
 }
 
 func GetAndMaybeCacheTreeFromRootDirectoryDigest(ctx context.Context, casClient repb.ContentAddressableStorageClient, r *digest.CASResourceName, fc interfaces.FileCache, bs bspb.ByteStreamClient) (*repb.Tree, error) {
+	return retry.Do(ctx, retryOptions("GetTree"), func(ctx context.Context) (*repb.Tree, error) {
+		return getAndMaybeCacheTreeFromRootDirectoryDigest(ctx, casClient, r, fc, bs)
+	})
+}
+
+func getAndMaybeCacheTreeFromRootDirectoryDigest(ctx context.Context, casClient repb.ContentAddressableStorageClient, r *digest.CASResourceName, fc interfaces.FileCache, bs bspb.ByteStreamClient) (*repb.Tree, error) {
 	var dirs []*repb.Directory
 	if fc != nil && bs != nil && *requestCachedSubtreeDigests {
 		out, err := getAndCacheTreeFromRootDirectoryDigest(ctx, casClient, r, fc, bs)
 		if err != nil {
+			if status.IsNotFoundError(err) {
+				return nil, retry.NonRetryableError(err)
+			}
 			return nil, err
 		}
 		dirs = out
 	} else {
 		out, subTrees, err := streamTree(ctx, casClient, r, false)
```

Comment:
looking at the code a little closer, i think `streamTree` is the only call that does not have retries, the other code paths have RPC-level retries so I don't know if it's worth retrying the whole thing again on top of the retries already having been done at the RPC level. I think it would be better if we added retries to the RPC in `streamTree`.

Reply:
Sounds good, moved the retries into there.